engines/e_pkcs11.c

```diff
@@ -0,0 +1,490 @@
+        PKCS11_trace("C_FindObjects failed, error: %#04X\n", rv);
+        PKCS11err(PKCS11_F_PKCS11_GET_PRIVATE_KEY,
+                  PKCS11_R_FIND_OBJECT_FAILED);
+        goto err;
+    }
+    return objhandle;
+
+ err:
+    return -1;
+}
+
+char pkcs11_hex_int(char nib1, char nib2)
+{
+    int ret = (nib1-(nib1 <= 57 ? 48 : (nib1 < 97 ? 55 : 87)))*16;
+    ret += (nib2-(nib2 <= 57 ? 48 : (nib2 < 97 ? 55 : 87)));
+    return ret;
+}
+
+int pkcs11_parse_uri(const char *path, char *token, char **value)
+{
+    char *tmp, *end, *hex2bin;
+    int j = 0;
+    if ((tmp = strstr(path, token)) == NULL)
+        return 0;
+    tmp += strlen(token);
+    *value = malloc(strlen(tmp) + 1);
+    end = strpbrk(tmp, ";");
+    snprintf(*value, end == NULL ? strlen(tmp) + 1 : end - tmp + 1, "%s", tmp);
+    hex2bin = malloc(strlen(*value) + 1);
+    for (int i = 0; i < strlen(*value); i++) {
+        if (*(*value+i) == '%' && i < (strlen(*value)-2)) {
+            *(hex2bin+j) = pkcs11_hex_int(*(*value+i+1), *(*value+i+2));
+            i += 2;
+        } else {
+            *(hex2bin+j) = *(*value+i);
+        }
+        j++;
+    }
+    *(hex2bin+j) = '\0';
+    *value = hex2bin;
+    return 1;
+}
+
+static EVP_PKEY *pkcs11_engine_load_private_key(ENGINE * e, const char *path,
+                                                UI_METHOD * ui_method,
+                                                void *callback_data)
+{
+    CK_ULONG kt, class;
+    CK_ATTRIBUTE key_type[] = {
+        { CKA_CLASS,    &class, sizeof(class) },
+        { CKA_KEY_TYPE, &kt,    sizeof(kt)    }
+    };
+    EVP_PKEY *k = NULL;
+    CK_RV rv;
+    char *id, *pin;
+    char *module_path;
+
+    if (!strncmp(path, "pkcs11:", 7)) {
+        path += 7;
+
+	if (!pkcs11_parse_uri(path,"module-path=", &module_path))
+           goto err;
+	if (!pkcs11_parse_uri(path,"id=", &id))
+           goto err;
+	if (!pkcs11_parse_uri(path,"pin-value=", &pin))
+           goto err;
+
+    } else {
+        PKCS11_trace("String pkcs11: not found\n");
+        goto err;
+    }
+
+    pkcs11st.id = (CK_BYTE *) id;
+    pkcs11st.pin = (CK_BYTE *) pin;
+
+    rv = pkcs11_initialize(module_path);
+
+    if (rv != CKR_OK) {
+        goto err;
+    }
+
+    pkcs11st.session = pkcs11_start_session(pkcs11_get_slot());
+    pkcs11_login(pkcs11st.session, pkcs11st.pin, CKU_USER);
+    pkcs11st.key = pkcs11_get_private_key(pkcs11st.session, pkcs11st.id,
+                                          strlen((char *)pkcs11st.id));
+
+    rv = pkcs11_funcs->C_GetAttributeValue(pkcs11st.session,
+                                    pkcs11st.key, key_type, 2);
+
+    if(rv != CKR_OK || class != CKO_PRIVATE_KEY) {
+        return k;
+    }
+
+    if(kt == CKK_RSA) {
+        CK_ATTRIBUTE rsa_attributes[] = {
+            { CKA_MODULUS, NULL, 0 },
+            { CKA_PUBLIC_EXPONENT, NULL, 0 }
+        };
+        RSA *rsa = RSA_new();
+
+        if(((rv = pkcs11_funcs->C_GetAttributeValue(pkcs11st.session,
```

Comment:
I really don't like this as a single long if statement - it makes it harder to read.
It is a better style to put the get+alloc+get into a service function that you call here.